src/languageservice/services/yamlRename.ts

```diff
@@ -0,0 +1,201 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Red Hat, Inc. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { TextDocument } from 'vscode-languageserver-textdocument';
+import { Position, Range, WorkspaceEdit, TextEdit } from 'vscode-languageserver-types';
+import { yamlDocumentsCache } from '../parser/yaml-documents';
+import { matchOffsetToDocument } from '../utils/arrUtils';
+import { TextBuffer } from '../utils/textBuffer';
+import { Telemetry } from '../telemetry';
+import { CST, isAlias, isCollection, isScalar, visit, Node } from 'yaml';
+import { SourceToken, CollectionItem } from 'yaml/dist/parse/cst';
+import { SingleYAMLDocument } from '../parser/yamlParser07';
+import { isCollectionItem } from '../utils/astUtils';
+import { PrepareRenameParams, RenameParams } from 'vscode-languageserver-protocol';
+
+interface RenameTarget {
+  anchorNode: Node;
+  token: SourceToken;
+  yamlDoc: SingleYAMLDocument;
+}
+
+export class YamlRename {
+  constructor(private readonly telemetry?: Telemetry) {}
+
+  prepareRename(document: TextDocument, params: PrepareRenameParams): Range | null {
+    try {
+      const target = this.findTarget(document, params.position);
+      if (!target) {
+        return null;
+      }
+      if (!this.findAnchorToken(target.yamlDoc, target.anchorNode)) {
+        return null;
+      }
+      return this.getNameRange(document, target.token);
+    } catch (err) {
+      this.telemetry?.sendError('yaml.prepareRename.error', err);
+      return null;
+    }
+  }
+
+  doRename(document: TextDocument, params: RenameParams): WorkspaceEdit | null {
+    try {
```

Comment:
Currently with this PR, it's possible to rename the anchor to a name that's invalid (for instance, add any of `[`, `]`, `(`, `)`, `,`). For a list of what characters are valid for an anchor, here's what the specification says: https://yaml.org/spec/1.2.2/#rule-ns-anchor-char.

There's a way to reject these names. Are you interested in working on this? I think it's important to have, but it's a bit more work, and as it stands this PR is helpful as-is.